alacritty/src/cursor.rs

```diff
@@ -1,112 +1,79 @@
-//! Helpers for creating different cursor glyphs from font metrics.
+//! Helpers for creating different cursors with rects.
+use alacritty_terminal::ansi::CursorShape;
+use alacritty_terminal::term::color::Rgb;
+use alacritty_terminal::term::{CursorKey, RenderableCursor, SizeInfo};
 
-use crossfont::{BitmapBuffer, Metrics, RasterizedGlyph};
+use crate::renderer::rects::RenderRect;
 
-use alacritty_terminal::ansi::CursorShape;
+struct Cursor {
+    /// Coordinates of the top left corner of the cursor cell.
+    pub x: f32,
+    pub y: f32,
 
-pub fn get_cursor_glyph(
-    cursor: CursorShape,
-    metrics: Metrics,
-    offset_x: i8,
-    offset_y: i8,
-    is_wide: bool,
-    cursor_thickness: f64,
-) -> RasterizedGlyph {
-    // Calculate the cell metrics.
-    //
-    // NOTE: With Rust 1.47+ `f64 as usize` is defined to clamp automatically:
-    // https://github.com/rust-lang/rust/commit/14d608f1d8a0b84da5f3bccecb3efb3d35f980dc
-    let height = (metrics.line_height + f64::from(offset_y)).max(1.) as usize;
-    let mut width = (metrics.average_advance + f64::from(offset_x)).max(1.) as usize;
-    let line_width = (cursor_thickness * width as f64).round().max(1.) as usize;
-
-    // Double the cursor width if it's above a double-width glyph.
-    if is_wide {
-        width *= 2;
-    }
+    pub color: Rgb,
 
-    match cursor {
-        CursorShape::HollowBlock => get_box_cursor_glyph(height, width, line_width),
-        CursorShape::Underline => get_underline_cursor_glyph(width, line_width),
-        CursorShape::Beam => get_beam_cursor_glyph(height, line_width),
-        CursorShape::Block => get_block_cursor_glyph(height, width),
-        CursorShape::Hidden => RasterizedGlyph::default(),
-    }
+    pub key: CursorKey,
+
+    pub thickness: f64,
 }
 
-/// Return a custom underline cursor character.
-pub fn get_underline_cursor_glyph(width: usize, line_width: usize) -> RasterizedGlyph {
-    // Create a new rectangle, the height is relative to the font width.
-    let buf = vec![255u8; width * line_width * 3];
-
-    // Create a custom glyph with the rectangle data attached to it.
-    RasterizedGlyph {
-        c: ' ',
-        top: line_width as i32,
-        left: 0,
-        height: line_width as i32,
-        width: width as i32,
-        buf: BitmapBuffer::RGB(buf),
+impl Cursor {
+    fn new(cursor: RenderableCursor, size_info: &SizeInfo, thickness: f64) -> Self {
```

Comment:
I find this API a bit strange, how you create a cursor first and then create a cursor rect from it.

You should know what kind of cursor you have from the beginning, right? So why not just create one immediately?

You also do strange stuff like creating a cursor, passing `size_info` to it, but then getting the `width` out of size info, multiplying it by the width the created cursor gives you, just to pass it back to the cursor. After you passed the `size_info` to the cursor it had all the information it ever needed, why juggle around with the `width` outside of it?